Improve log output for delete cluster

oxr463 opened the pull request:
- [x] Deleting node
- [x] Deleting kubeconfig

Log output:

```
kind delete cluster -n kind
Deleting cluster "kind" ...
Removed context "kind" from kubeconfig
Deleted nodes: ["kind-control-plane"]
```

Closes: https://github.com/kubernetes-sigs/kind/issues/2924

BenTheElder commented on pkg/cmd/kind/delete/cluster/deletecluster.go:
```diff
@@ -74,5 +74,6 @@ func deleteCluster(logger log.Logger, flags *flagpole) error {
 	if err := provider.Delete(flags.Name, flags.Kubeconfig); err != nil {
 		return errors.Wrapf(err, "failed to delete cluster %q", flags.Name)
 	}
+	logger.V(0).Infof("Deleted cluster: %q", flags.Name)
```
Sorry, but this is not what we discussed adding in #2924.

We should log when we take specific actions like deleting node containers or if the kubeconfig is modified.
This long line doesn't add any additional information versus the log line above on line 73.

oxr463 replied:
I've amended my PR. I don't think I can log the deletion of specific nodes unless I modify the `Provider.DeleteNodes()` which doesn't seem to have access to the logging interface.

Reference(s):
- https://github.com/kubernetes-sigs/kind/blob/main/pkg/cluster/internal/providers/provider.go#L41

oxr463 commented:
The error handling doesn't seem to catch deleting non-existing clusters:

```
kind get clusters
No kind clusters found.

kind delete cluster -n non-existent
Deleting cluster "non-existent" ...
Removing context "non-existent" from kubeconfig ...
Deleting nodes for cluster "non-existent" ...
```

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,16 +33,17 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
 		return errors.Wrap(err, "error listing nodes")
 	}
 
+	logger.V(0).Infof("Removing context %q from kubeconfig ...", name)
```
The call below may actually do nothing. We should only log more info when we actually did something.

oxr463 replied:
I've moved this into an else clause.

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,16 +33,17 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
 	kerr := kubeconfig.Remove(name, explicitKubeconfigPath)
 	if kerr != nil {
 		logger.Errorf("failed to update kubeconfig: %v", kerr)
+		return err
```
This is wrong

oxr463 replied:
Fixed.

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,16 +33,17 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
-		return err
-	}
-	if kerr != nil {
+		logger.Errorf("failed to delete nodes: %v", kerr)
```
This is redundant to returning the error.

oxr463 replied:
Removed.

BenTheElder replied:
still redundant.

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,16 +33,17 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
+		return err
 	}
 
+	logger.V(0).Infof("Deleting nodes for cluster %q ...", name)
```
This doesn't tell the user if we actually found any containers and removed them.

BenTheElder commented:
> The error handling doesn't seem to catch deleting non-existing clusters:

From the past discussion in the issues ... that is working as intended.

there's no such thing as a cluster existing.

There is however node containers associated with a cluster name or kubeconfig entries.

Deleting cluster resources is reentrant, like `rm`, it makes it easier to script reliably. If the command fails in removing nodes that exist it will exit non-zero and log an error. Otherwise it will not and you can have multiple cleanup paths attempt to call it.

oxr463 commented:
```
kind delete clusters -A
Removed context "bar" from kubeconfig
Removed context "foo" from kubeconfig
Deleted clusters: ["bar" "foo"]
```

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -36,13 +36,14 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
 		logger.Errorf("failed to update kubeconfig: %v", kerr)
+		return kerr
+	} else {
+		logger.V(0).Infof("Removed context %q from kubeconfig", name)
```
this will tell the user we removed it from the kubeconfig even if the kubeconfig didn't contain the context. the context name is also not just the cluster name, and the context is not the only thing in the kubeconfig.

oxr463 replied:
Used context name instead of cluster name.

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -36,13 +36,14 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
+		logger.V(0).Infof("Removed context %q from kubeconfig", name)
 	}
 
 	err = p.DeleteNodes(n)
```
we know here that we're deleting nodes n, we can log those if there are any.

oxr463 commented:
```
kind delete clusters -A
Removed context "kind-bar" from kubeconfig
Deleted nodes: ["bar-control-plane"]
Removed context "kind-foo" from kubeconfig
Deleted nodes: ["foo-control-plane"]
Deleted clusters: ["bar" "foo"]
```

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -35,15 +35,16 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
-		logger.Errorf("failed to update kubeconfig: %v", kerr)
+		return kerr
+	} else {
+		logger.V(0).Infof("Removed context %q from kubeconfig", kubeconfig.ContextForCluster(name))
```
this isn't strictly true, the context may not have been present in the cluster

this probably is most easily accomplished inside Remove

oxr463 replied:
Remove just returns RemoveKIND, which doesn't seem to have access to the logger.

oxr463 replied:
@BenTheElder what do you think about modifying RemoveKIND() to add the logger as a parameter?

oxr463 replied:
Ping @BenTheElder

BenTheElder replied:
alternatively we could return true/false if the cluster is removed or not.

I think it would be reasonable to add the logger as a parameter.

BenTheElder commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -35,15 +35,16 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
-	if kerr != nil {
-		return err
+	} else {
+		logger.V(0).Infof("Deleted nodes: %q", n)
```
We shouldn't log this when len(nodes) = 0

oxr463 replied:
How can I test a cluster with no nodes?

oxr463 replied:
@BenTheElder I don't see how this would ever trigger.

oxr463 replied:
Ping @BenTheElder

BenTheElder replied:
this is a public API, so anyone can call it against this for a cluster name even if there are no nodes

BenTheElder replied:
specifically by calling https://github.com/kubernetes-sigs/kind/blob/d53225750af998c6218b5c8c730b8b3ca0e2d472/pkg/cluster/provider.go#L186-L188

BenTheElder replied:
we can check len(nodes) and avoid the deleteNodes call and the logging when there are no nodes

oxr463 commented:
@BenTheElder ping 🙂

stmcginnis commented:
Looks like some tests need to be updated:

```txt
pkg/cluster/internal/kubeconfig/internal/kubeconfig/remove_test.go:186:30: not enough arguments in call to RemoveKIND
	have (string, string)
	want (log.Logger, string, string)
```

oxr463 commented:
/retest

oxr463 commented:
@stmcginnis any ideas why there are issues with the CSI driver and volumes? I don't think that's related to my changes.

stmcginnis commented:
> @stmcginnis any ideas why there are issues with the CSI driver and volumes? I don't think that's related to my changes.

I'm not sure, but I agree that doesn't look like it could be related to your changes. It might just be a transient failure.

/retest

oxr463 commented:
/retest

oxr463 commented:
@stmcginnis can you please add the label for tide?

aojea commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,17 +33,19 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
+	kerr := kubeconfig.Remove(logger, name, explicitKubeconfigPath)
 	if kerr != nil {
 		logger.Errorf("failed to update kubeconfig: %v", kerr)
+		return kerr
```
@BenTheElder we don't try to delete the nodes anymore?

stmcginnis replied:
That's a good point. While it's not great if kubeconfig can't be updated, we probably shouldn't just error out completely at that point.

I could see either holding on to `kerr` and adding one more check after the `DeleteNodes` block is complete:

```go
if kerr != nil {
    return kerr
}

return nil
```

Or just ignoring it and going with just the logging of the error.

The current code uses the former approach. We should probably stick with that.

oxr463 replied:
Is there anything I need to change or is this good enough as is?

If you look through the history of this PR, I had made some changes to this but I ended up putting it back.

stmcginnis replied:
I think I would prefer the change I noted above. I think that's what @aojea was alluding to as well.

oxr463 replied:
What about the `return nil` down on line 50?

oxr463 replied:
Maybe I don't understand. That's already in place.

stmcginnis replied:
- [ ] Delete line 39
- [ ] Insert the following before the `return nil` at the end:

```go
if kerr != nil {
    return kerr
}
```

oxr463 replied:
Alright done. But what about the `return err` on line 44?

stmcginnis replied:
True, otherwise if their is a failure removing the kubeconfig, then the node deletion fails, they never see that. Though then they would still need to fix the issue and clean things up anyway.

Maybe it would be better to move the kubeconfig deletion completely after the node deletion? That way if the kubeconfig deletion fails, at least the cluster itself has already been removed and the user would just need to clean up their kubeconfig.

I think either way is fine. Let's see if others have a stronger opinion on this.

Otherwise looks great - thanks for the updates!

oxr463 replied:
@aojea @BenTheElder any updates here?

BenTheElder commented on pkg/cluster/internal/kubeconfig/internal/kubeconfig/remove.go:
```diff
@@ -48,6 +49,7 @@ func RemoveKIND(kindClusterName string, explicitPath string) error {
 				if err := write(existing, configPath); err != nil {
 					return err
 				}
+				logger.V(0).Infof("Removed context %q from kubeconfig", kindClusterName)
```
We should note this and log it later, we should not log while holding locks, the logging is comparatively much slower than the other operations.

oxr463 replied:
How should I go about that?

BenTheElder replied:
for what we're logging here currently we could toggle a boolean and then log based on that boolean after the logs are released

BenTheElder commented on pkg/cluster/internal/kubeconfig/internal/kubeconfig/remove_test.go:
```diff
@@ -183,7 +184,7 @@ users:
 
 	// ensure that we can write this merged config
-	if err := RemoveKIND("foo", existingConfigPath); err != nil {
+	if err := RemoveKIND(cmd.NewLogger(), "foo", existingConfigPath); err != nil {
```
We should be testing with a fake. There is a Noop logger in the log package.

BenTheElder reviewed (commented):
sorry, most of my energy in this project is still going to bug fixes and maintenance, elsewhere it's been quite a messy time with the kubernetes project almost running out of infra funding and other fun 🙃

stmcginnis commented:
/ok-to-test
/lgtm

Looks like all comments have been addressed. This looks good to me. Thanks!

aojea commented on pkg/cluster/internal/kubeconfig/internal/kubeconfig/remove.go:
```diff
@@ -52,10 +52,10 @@ func RemoveKIND(kindClusterName string, explicitPath string) error {
 			return nil
 		}(configPath); err != nil {
-			return err
+			return false, err
```
is there any situation that the bool is not false if err == nil ?

aojea commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -33,17 +33,25 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
 	}
 
-	kerr := kubeconfig.Remove(name, explicitKubeconfigPath)
+	l, kerr := kubeconfig.Remove(name, explicitKubeconfigPath)
+	if l {
+		logger.V(0).Infof("Removed context %q from kubeconfig", name)
+	}
```
do we really need this?

oxr463 replied:
This is what was asked for by @BenTheElder

oxr463 replied:
I’m fine to take it out if that’s what we want to do.

oxr463 replied:
Alright, I removed it.

aojea commented on pkg/cluster/internal/delete/delete.go:
```diff
@@ -38,12 +38,17 @@ func Cluster(logger log.Logger, p providers.Provider, name, explicitKubeconfigPa
+
 	if kerr != nil {
-		return err
+		return kerr
```
heh, you also fixed this bug

aojea commented:
/lgtm
/approve

BenTheElder commented:
This is released, please see the note at https://github.com/kubernetes-sigs/kind/releases/tag/v0.18.0#contributors and the "new features section" @oxr463
